File: 2-2.lua
#!/usr/bin/env lua5.4

local a = 0
local x = 0
local d = 0

local env = {
	forward = function(n) x = x + n; d = d + a*n end,
	up = function(n) a = a - n end,
	down = function(n) a = a + n end,
}

for l in io.lines("2.txt") do
	load(l:gsub(' ', '('):gsub('$', ')'), "2.txt", "t", env)()
end

print(x*d)

File: 2-2.hs
#!/usr/bin/runghc

import Data.Char

data Instr = Forward Int
           | Down    Int
           | Up      Int
    deriving (Read)

readInstr :: String -> Instr
readInstr []     = undefined
readInstr (x:xs) = read $ toUpper x : xs

move :: (Int, Int, Int) -> Instr -> (Int, Int, Int)
move (a, x, d) (Forward n) = (a    , x + n, d + a*n)
move (a, x, d) (Down    n) = (a + n, x    , d      )
move (a, x, d) (Up      n) = (a - n, x    , d      )

main :: IO ()
main = do
    instrs <- fmap (map readInstr . lines) $ readFile "2.txt"
    let pos@(_, x, d) = foldl move (0, 0, 0) instrs
    print pos
    print $ x*d
